### ISSUE TITLE
Proposal: Schedule learning rate and other parameters based on variable

### ISSUE THREAD

**Issue by lucasdavid** [keras-team/keras#16696]
**Describe the feature and the current behavior/state**.
The current schedules api has a few drawbacks:
- The `LearningRateSchedule.__call__(step: int)` is invoked at the beginning of `apply_gradients`. Its result is stored in `opt._current_learning_rate: tf.Variable` and shared amongst all variables being optimized;
- Hyper-parameters that should be jointly scheduled with `lr` are not (e.g.: `experimental.AdamW.weight_decay`);
- It's impossible to schedule the `lr` based on which variable is being updated (e.g.: `lr=0.1` for scratch weights and `lr=0.01` for pretrained weights; or apply weight decay for kernels, but not for biases and batch norm layers), which forces us to use the much more complex `tfa.optimizers.MultiOptimizer`;
- Accessing the learning rate inside the optimizer is a little ugly, as you need to check if its a `float`, function or a `Schedule`.

**Proposal**
1. To abstract the `LearningRateSchedule` class onto the `Schedule` (or `VariableSchedule`) class and allow it to be used in any context. Examples:
    - `AdamW(learning_rate=0.001, weight_decay=ExponentialDecay(0.001, max_steps, 1e-5))`;
    - `keras_cv.layers.DropBlock2D(rate=LinearIncrease(initial_value=0.2, rate=1e-5, max_value=.9)))`
2. To add an optional parameter `variable` in the `Schedule` call (`Schedule.__call__(step: int, variable: str = None)` and make a call **for each iteration and variable**, instead of each iteration;
3. Create a default schedule `schedules.Constant(Schedule)` and always construct it when passing constant parameters (learning rate, weight decay). Clean up optimizers' code to always defer to `lr(self.iterations)` or `lr(self.iterations, var_key)`;
4. Remove fields such as `AdamW#_exclude_from_weight_decay` and other snippets that condition behavior based on the observed variable. The users can instead sub-class `Schedule` and write the logic that ignore the weights they want.

**Will this change the current api? How?**
The public API won't be affected, just the code inside the optimizers. These classes can be simplified (remove type checks for learning_rate) and any access to the learning rate should be a call with `self.iterations` and the current `var_key` (if available).

**Who will benefit from this feature?**
- Users who want different learning rates for different variables (example 1)
- Users who wish to limit the application of decoupled weight decay to specific variables (example 1)
- Users who want to schedule other factors in their own components (such as parameters in Layers, example 2)

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): Yes
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

```py
class AdamW(optimizer.Optimizer):
    def update_step(self, gradient, variable):
        """Update step given gradient and the associated model variable."""
        beta_1_power = None
        beta_2_power = None
        # lr = tf.cast(self.learning_rate, variable.dtype)  # (removed)
        local_step = tf.cast(self.iterations + 1, variable.dtype)
        beta_1_power = tf.pow(tf.cast(self.beta_1, variable.dtype), local_step)
        beta_2_power = tf.pow(tf.cast(self.beta_2, variable.dtype), local_step)

        var_key = self._var_key(variable)
        lr = tf.cast(self._learning_rate(self.iterations, var_key), variable.dtype)  # (added) conditioned to `var_key`
        m = self._momentums[self._index_dict[var_key]]
        v = self._velocities[self._index_dict[var_key]]
        ...
```
**Example 1**
```py
class Poly(Schedule):
  def __init__(self, default_value, params):
    self.default_value = default_value
    self.params = params

  def __call__(self, step, variable: str = None):
    if variable is not None:
      for name, value in self.params.items():
        if re.search(name, variable):
          return value
    return self.default_value

opt = SGD(
  learning_rate=Poly(.01, {
    'head/kernel': 0.2,
    'head/bias': 0.1,
    'backbone/conv.*/kernel': 0.02,
    'backbone/conv.*/bias': 0.01
  }),
  weight_decay=Poly(.0, {'head/kernel': 1e-4, 'backbone/conv.*/kernel': 1e-5})
)
```
**Example 2**
```py
epochs = 10
train_batches = tfds.load(...)[0].batch(32).prefetch(tf.data.AUTOTUNE)
max_steps = len(train_data) * epochs
opt = SGD()

class LinearIncrease(Schedule):
  def __init__(self, rate, initial_value=0., max_value=1.):
    self.rate = rate
    self.initial_value = initial_value
    self.max_value = max_value

  def __call__(self, step, variable: str = None):
    rate = self.rate*step
    current = self.initial_value*rate
    return tf.minimum(current, self.max_value)

def block(x, dropblock_rate=0.0, iterations=None):
  x = Conv2D(...)(x)
  x = BatchNormalization()(x)
  x = Activation(...)(x)
  x = DropBlock2D(dropout_rate=dropblock_rate)(x, iterations=iterations)

db_r_schedule = LinearIncrease(rate=2/max_steps, initial_value=0.2, max_value=.9)
  
y = x = tf.keras.Input([None, None, 3])
for i in range(3):
  y = block(y, db_r_schedule, opt.iterations)
model = Model(x, y)
model.compile(optimizer=opt, ...)
model.fit(...)
```

**Comment by chenmoneygithub** [comment]
@lucasdavid Thanks for reporting this issue!

There are two parts in this issue:
1. A `Schedule` class that can work for our different layers. We think this is important, and will make some design for it.
2. Per variable learning rate. My main question for this is it widely used? or in what situations do we use it? Thanks!

**Comment by lucasdavid** [comment]
There are two parts, but I think they are strongly coupled. Maybe my examples weren't very good, so just to make sure we are on the same page: Is it your understanding that the `Schedule` should work for different hyperparameters (not only for different layers), including other parameters inside the `Optimizer` (e.g. momentum, weight decay)? I'm not proposing different learning rates for different variables, but **different hyperparameter values** for different variables.
This resembles the parameter groups feature implemented in [PyTorch's optim api](https://pytorch.org/docs/stable/optim.html#per-parameter-options).

Different parameter groups are useful to increase speed of training, to isolate behavior for certain variables and to bring stability to the training procedure. Specifically for the learning rate, it quickly updates scratch weights to the same magnitude of the pretrained weights. For weight decay, it can isolate variables that should not suffer decay (e.g. biases, batchnorm's gamma and beta). It could also be used to isolate scheduling logic/acceleration strategies to specific parts of the model. For example: the classification head is trained with nesterov momentum with constant lr while the backbone is trained with SGD with Exponential Decay.

The examples of this that I know of are:
- [OC-CSE/models/model_cse.py#L101](https://github.com/KAIST-vilab/OC-CSE/blob/main/models/model_cse.py#L101)
- [Puzzle-CAM/train_classification.py#L188](https://github.com/OFRIN/PuzzleCAM/blob/master/train_classification.py#L188)
- [psa/train_aff.py#L69](https://github.com/jiwoon-ahn/psa/blob/master/train_aff.py#L69)
- [ResNeSt/scripts/torch/train.py](https://github.com/zhanghang1989/ResNeSt/blob/master/scripts/torch/train.py#L162)
- [ResNet38-Semantic-Segmentation/train_seg.py#L50](https://github.com/peihan-miao/ResNet38-Semantic-Segmentation/blob/master/train_seg.py#L50)

**Comment by lucasdavid** [comment]
Ok, great. Let me know if you are interested in the `variable` parameter in `LearningRateSchedule#__call__` as well. I believe this adds value by easing the replication of PyTorch's models, but if you decide it's out of the scope of this repo, I'll close the issue.
